udev rules to prioritize network interfaces

macmpi opened the pull request:
This PR replaces [Pi3-specific driver blacklist-based filter](https://github.com/volumio/Build/issues/150) by udev rule.
It brings several benefits:
- avoids unwanted brcmfmac/brcmutil driver loads on all Pi platforms
- prioritizes plugged-in external USB network interfaces over built-in (wlan & eth)
- generic for all platforms (+added customization for Pi built-in eth which **is** on USB)

This fix has been tested with several devices and dongles.

volumio commented:
That's clever. Has it been tested extensively? (trying to avoid regressions)

macmpi commented:
Thanks.

Extensively, not really as I do not own sufficient HW variables.
I made many unitary tests for base search on my PiZero with my USB dongles (wifi 8188, ethernet asix), and noted no issue.
The only test on Pi3 is on the forum thread: only wifi part has been validated with a realtek dongle.
You may want to check with your Pi3 (seems you have one) and wifi dongles.
If you have USB-ethernet dongle, I'm interested in your feedback too.

If you experience issues, please send me **udevadm info /sys/class/net/YOUR_INTERFACE** so that I can eventually fine-tune the filters.

chsims1 commented:
I have a Pi3 with several USB wifi adapters. Exactly what do you need testing?  Simply udev rule inclusion and look at interfaces brought up?

macmpi commented:
Sure, that would be great, thanks!

Indeed, you have to restart with/without USB adapter plugged, and check which interface is actually used by Volumio.
You can confirm that's the one you want, by getting current interface details with `udevadm info /sys/class/net/wlan0` (or eth0).

You'll also want to (make copies elsewhere first):
- `sudo rm /etc/modprobe.d/raspi-blacklist.conf`
- keep only last line of /bin/wifistart.sh
This will replicate the complete change

Looking forward your feedback.

chsims1 commented:
RTL8188cus: works as wanted, USB dongle wlan0, RPi internal wlan1
RTL8192cu: USB dongle 'rename4', RPi internal wlan0
RT5370: USB dongle 'rename4', RPi internal wlan0
Atheros (AR9170+AR9101): USB dongle didn't work, RPi internal wlan0

macmpi commented:
ok, thanks.
Could you please provide output of **udevadm info /sys/class/net/YOUR_INTERFACE** on those failing please? (maybe on pastebin)?

For few (rename4 in particular) it may be because of existing ralink rule installed by mrEngman drivers: could you `sudo rm /etc/udev/rules.d/95-ralink.rules` (keep a copy before) and check if it has any effect?
Also I noted that base `/lib/udev/rules.d/75-persistent-net-generator.rules` differ from Raspian's and may have effect.

chsims1 commented:
Well, on second time of asking RTL8192cu comes up as wlan0 (RPi internal wlan1). Strange??

RT5370:

`volumio@volumiokitchen:~$ udevadm info /sys/class/net/rename4                                                                                                
P: /devices/platform/soc/3f980000.usb/usb1/1-1/1-1.4/1-1.4:1.0/net/rename4                                                                                   
E: DEVPATH=/devices/platform/soc/3f980000.usb/usb1/1-1/1-1.4/1-1.4:1.0/net/rename4                                                                           
E: DEVTYPE=wlan                                                                                                                                              
E: ID_BUS=usb                                                                                                                                                
E: ID_MODEL=802.11_n_WLAN                                                                                                                                    
E: ID_MODEL_ENC=802.11\x20n\x20WLAN                                                                                                                          
E: ID_MODEL_FROM_DATABASE=RT5370 Wireless Adapter                                                                                                            
E: ID_MODEL_ID=5370
E: ID_NET_DRIVER=rt2800usb
E: ID_NET_NAME_MAC=wlx7cdd90b0dec5
E: ID_OUI_FROM_DATABASE=Shenzhen Ogemray Technology Co., Ltd.
E: ID_PATH=platform-3f980000.usb-usb-0:1.4:1.0
E: ID_PATH_TAG=platform-3f980000_usb-usb-0_1_4_1_0
E: ID_REVISION=0101
E: ID_SERIAL=Ralink_802.11_n_WLAN_1.0
E: ID_SERIAL_SHORT=1.0
E: ID_TYPE=generic
E: ID_USB_DRIVER=rt2800usb
E: ID_USB_INTERFACES=:ffffff:
E: ID_USB_INTERFACE_NUM=00
E: ID_VENDOR=Ralink
E: ID_VENDOR_ENC=Ralink
E: ID_VENDOR_FROM_DATABASE=Ralink Technology, Corp.
E: ID_VENDOR_ID=148f
E: IFINDEX=4
E: INTERFACE=rename4
E: SUBSYSTEM=net
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/rename4
E: TAGS=:systemd:
E: USEC_INITIALIZED=66025`

Removing the ralink rule made no difference.

macmpi commented:
let me look...
for Atheros, it's probably missing firmware: you may want to `sudo apt-get install firmware-linux-free`

macmpi commented:
Can you please make sure `/etc/udev/rules.d/70-persistent-net.rules` is this:
```
ACTION=="add", SUBSYSTEM=="net", SUBSYSTEMS=="usb", KERNEL=="wlan*", ATTR{type}=="1", NAME="wlan0"
ACTION=="add", SUBSYSTEM=="net", SUBSYSTEMS=="usb", KERNEL=="eth*", DRIVERS!="smsc95*", ATTR{type}=="1", NAME="eth0"
```
`/etc/udev/rules.d/95-ralink.rules` does not exist anymore
And you do a cold reboot each time to change things (**removing power**).

With that, are RTL8192cu and RT5370 still "faulty" in their names?

volumio commented:
before removing power, make sure you type: 
sync

macmpi commented:
yes remove power after proper shutdown indeed!

chsims1 commented:
Ok, sorry for delay ... dog walking and making dinner got in the way ;).

Right, repeated for 4 dongles, ralink rule renamed, 70-persistent-net.rules as per your post above, sync & cold reboots between each.

RT8188cus: wlan0 and RPi internal on wlan1
RT9192cu: wlan0 and RPi internal on wlan1
RT5370: rename4 and RPi internal on wlan0
(AR9170+AR9101): rename4 and RPi internal on wlan0 (driver updated as above)

Don't know if it's useful but udevadm info /sys/class/net/rename4 for Atheros is:

volumio@volumiokitchen:~$ udevadm info /sys/class/net/rename4                                                                                                
P: /devices/platform/soc/3f980000.usb/usb1/1-1/1-1.4/1-1.4:1.0/net/rename4                                                                                   
E: DEVPATH=/devices/platform/soc/3f980000.usb/usb1/1-1/1-1.4/1-1.4:1.0/net/rename4                                                                           
E: DEVTYPE=wlan                                                                                                                                              
E: ID_BUS=usb                                                                                                                                                
E: ID_MODEL=USB2.0_WLAN                                                                                                                                      
E: ID_MODEL_ENC=USB2.0\x20WLAN                                                                                                                               
E: ID_MODEL_FROM_DATABASE=WN111(v2) RangeMax Next Wireless [Atheros AR9170+AR9101]                                                                           
E: ID_MODEL_ID=9001                                                                                                                                          
E: ID_NET_DRIVER=carl9170                                                                                                                                    
E: ID_NET_NAME_MAC=wlx00223f8b129f                                                                                                                           
E: ID_OUI_FROM_DATABASE=Netgear Inc.
E: ID_PATH=platform-3f980000.usb-usb-0:1.4:1.0
E: ID_PATH_TAG=platform-3f980000_usb-usb-0_1_4_1_0
E: ID_REVISION=0106
E: ID_SERIAL=ATHER_USB2.0_WLAN_12345
E: ID_SERIAL_SHORT=12345
E: ID_TYPE=generic
E: ID_USB_CLASS_FROM_DATABASE=Vendor Specific Class
E: ID_USB_DRIVER=carl9170
E: ID_USB_INTERFACES=:ff0000:
E: ID_USB_INTERFACE_NUM=00
E: ID_USB_PROTOCOL_FROM_DATABASE=Vendor Specific Protocol
E: ID_USB_SUBCLASS_FROM_DATABASE=Vendor Specific Subclass
E: ID_VENDOR=ATHER
E: ID_VENDOR_ENC=ATHER
E: ID_VENDOR_FROM_DATABASE=NetGear, Inc.
E: ID_VENDOR_ID=0846
E: IFINDEX=4
E: INTERFACE=rename4
E: SUBSYSTEM=net
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/rename4
E: TAGS=:systemd:
E: USEC_INITIALIZED=915012

macmpi commented:
> ralink rule renamed

renamed what? (names do matter for rules as they are executed in sequence): removed from that directory (or altogether) would be much safer.

chsims1 commented:
Ah sorry, I was under the impression that it was anything named .rules

macmpi commented:
Thanks for your kind help.
Could you look at `dmesg` for the early part of the startup process, if you see what happens with that interface: this may give clues. (put in pastebin if long log)

macmpi commented:
one other thing...can you comment (or remove) the second line (the one related to eth0) of the rule for now?
Does it change anything?  (including the rename#)?

chsims1 commented:
I've not truncated the output from dmesg: [http://pastebin.com/JzbazcyQ](url)

Removing the eth0 line from the rule made no apparent diffference.

udevadm info -a /sys/class/net/rename4: [http://pastebin.com/x4bqQjy5](url)

macmpi commented:
Thanks. Can you replace the rule content by this and tell if it helps?
```
ACTION=="add", SUBSYSTEM=="net", SUBSYSTEMS=="usb", ENV{DEVTYPE}=="wlan", NAME="wlan0", OPTIONS+="last_rule"
ACTION=="add", SUBSYSTEM=="net", SUBSYSTEMS=="usb", KERNEL=="eth*", DRIVERS!="smsc95*", NAME="eth0", OPTIONS+="last_rule"
```

chsims1 commented:
Dongle still named 'rename4'

macmpi commented:
`dmesg` has interesting bits, but indeed seems a bit shortened...can you get more through `journactl`?
rt2800usb seems to get to `wlan1` and then `rename4`
Is built-in still at `wlan0`? (dmesg seems to imply that name was already taken when bcm wanted it...or maybe it's due to a Volumio wifi restart)
Are other dongles fine with that rule?
(BTW do you have any ethernet-USB dongle too?)

chsims1 commented:
http://pastebin.com/rfFKqgxg

Built-in wifi still wlan0.

Just checking other dongles now.

Sorry, no ethernet-USB dongles

macmpi commented:
ok, [seems like](https://help.ubuntu.com/community/WifiDocs/Device/Tenda_W311M#Enable_support_in_udev) RT5370 needs extra config for proper udev support...

chsims1 commented:
I did cut&paste..now deleted and typed in ... no difference to naming behaviour.

Yes, the dongle drivers are working fine; I can connect to my home AP without problems (after disabling internal wifi).  However, neither Atheros or Ralink successfully created Volumio hotspot (indicated as 'on' in UI, but can't see network on my phone or a wifi analyser).

macmpi commented:
ok, thanks.
Will keep investigating then & try to fine-tune (have another idea, focusing the rule in opposite direction to target bcm chip instead): thanks again for all your helpful efforts.

Maybe @earlchew can bring expertise on overall `udev` science to help making it more "universal" to dongles.

chsims1 commented:
Good luck. Udev is very powerful, but a bit of a pig to fine-tune to achieve exactly what you want/expect.

macmpi commented:
@chsims1 the really amazing part is that your RT5370 uses the same driver as jncraig's Canakit dongle device that worked in [forum thread](https://volumio.org/forum/rpi-and-wifi-dongle-t6099-10.html#p29958): same udevadm info

macmpi commented:
may I dare asking testing this (big) one?
```
ACTION=="add", SUBSYSTEM=="net", SUBSYSTEMS!="usb", KERNEL=="wlan0", ATTR{type}=="1", ENV{internal_wlan0_exist}="true"
ACTION=="add", SUBSYSTEM=="net", SUBSYSTEMS=="usb", KERNEL=="wlan*", ATTR{type}=="1", ENV{external_wlan_exist}="true"
SUBSYSTEM=="net", SUBSYSTEMS!="usb", KERNEL=="wlan0", ATTR{type}=="1", ENV{external_wlan_exist}=="true", NAME="internal_wifi", OPTIONS+="log_priority=debug"
SUBSYSTEM=="net", SUBSYSTEMS=="usb", KERNEL=="wlan*", ATTR{type}=="1", ENV{internal_wlan0_exist}=="true", NAME="wlan0", OPTIONS+="log_priority=debug"
```
and eventually put `dmesg` in pastebin again?

chsims1 commented:
Well you've got rid of the 'rename4' :) ... RT5370 now appears as wlan1.

dmesg ....... http://pastebin.com/xRfCYw2W

macmpi commented:
and same, wlan1 for each others?

chsims1 commented:
RT5370: wlan1 (internal wlan0)
RTL8188cus: wlan0 (internal wlan1)
(AR9170 + AR9101): wlan1 (internal wlan0)
RTL8192cu: wlan0 (internal wlan1)

macmpi commented:
Ok. I'm going to try to remotely access a friend's Pi3 to directly test some base stuff on that HW next week (several things at play it seems). If you accept, I'll get back to you when I have something more reliable to test across dongles, not to abuse your time and patience. Thanks again!

volumio commented:
Still WIP?

macmpi commented:
Yes...planning a remote session with @chsims1 to figure-out specif issues with his dongles/setup.
Had things work on several Pi3, but I'd rather confirm with this one.
Will update title when done.

BTW, to you want me to keep empty `wifistart.sh` as empty placeholder, or shall I remove altogether?

volumio commented:
Let's keep it as an empty placeholder please (but I think we still need it to set power save off)

macmpi commented:
/etc/network/if-pre-up.d/powersavedisable does it already!

volumio commented:
Sometimes this does not work, for example for network devices turned on by wireless.js

macmpi commented:
@volumio said
> Sometimes this does not work, for example for network devices turned on by wireless.js

For my education, I'm curious which action in wireless.js does not relate to `ifupdown` and potentially omit `if-pre-up.d` scripts.
Quoting [this](http://askubuntu.com/questions/321979/when-does-ubuntu-12-04-run-network-if-pre-up-d-and-if-up-d-scripts#322654):
> About when scrips in if-pre-up.d and if-up.d are executed. They are only run for interfaces that are managed by ifupdown, that is they must be mentioned in /etc/network/interfaces. They are executed before and during setup of the interface that are to be started, as mentioned in the man page of interfaces.

I'll restore the line in `wifistart.sh` for precaution then.

As for the udev stuff: seems to work great on wlan side on @chsims1 setup: will let him confirm.
Still looking for eth0 side and tests... (Pi is special case here because built-in eth is on USB bus)

chsims1 commented:
Yep, all seems to be playing together nicely here.  Well done.

macmpi commented:
Glad it works!
Thanks so much for all your tests and kindness to provide remote access to get that done.

@volumio still giving a bit of thought & re-reading: will change PR title when ready for merge on my side.

macmpi commented:
@volumio I think it can be merged now.

volumio commented:
Ok, let me do last checks and then I'll merge

macmpi commented:
Please hold a bit pending a last test from @chsims1 with that rules:
```
ACTION=="add", SUBSYSTEM=="net", SUBSYSTEMS=="usb", ENV{DEVTYPE}=="wlan", ATTR{type}=="1", KERNEL=="wlan[1-9]*", NAME="wlan0"
ACTION=="add", SUBSYSTEM=="net", SUBSYSTEMS=="usb", ENV{DEVTYPE}!="wlan", ATTR{type}=="1", DRIVERS!="smsc95xx", KERNEL=="eth[1-9]*", NAME="eth0"
```

Simpler if no _rename4_ issues reported with that one (would not depend on extra code for naming).

macmpi commented:
Ok, so the whole current PR is needed to circumvent some potential race conditions in renaming interfaces with kernel's default names.
**So I consider PR "final"**: @volumio you may merge when you feel comfortable.

Longer term, it might be wise to consider naming Volumio network interfaces by some specific (and predictable) names, outside of kernel's default namespace (like _volumio_wlan_, _volumio_lan_ or any other better names).
One key benefit is that whatever kernel's naming conventions ([there could be at least 3](https://www.freedesktop.org/wiki/Software/systemd/PredictableNetworkInterfaceNames/)) across platforms' own upcoming kernel versions, Volumio interfaces would always be called in the same way (no bothering with kernel/systems updates).
From there, one could also think about a plugin to pick&choose preferred interface cards for Volumio, etc...while avoiding potential race conditions.
Just food for thoughts...not too big (albeit careful) changes needed in Volumio codebase.
